### PR TITLE
Add Conda package

### LINKED ISSUE #66: Create sbcl-librarian Conda package
## Notes
TODO

### PR DESCRIPTION
**Pull request by kartik-s**
Closes #66.

# Summary
This PR adds a Conda recipe for building a package called `sbcl-librarian` containing the following artifacts:
- `lib/libsbcl_librarian.so`
- `lib/sbcl_librarian.core`
- `include/sbcl_librarian.h`
- `include/sbcl_librarian_err.h`
- the `sbcl_librarian` Python package with wrappers for `libsbcl_librarian.so`

The CI workflows on all platforms are updated so that they:

1. Build SBCL + libsbcl from source
2. Build the `sbcl-librarian` Conda package and install it into an environment
3. Build `calc.dll` as a FASL-packed DLL against `libsbcl_librarian.dll` in that environment
4. Run the example C and Python programs against `calc.dll`

# Artifacts
## libsbcl_librarian.so
Built via the CMakeLists.txt file in `lib/`, which lists its sources and dependent libraries:

```cmake
add_library(sbcl_librarian SHARED sbcl_librarian.c sbcl_librarian.h sbcl_librarian_err.h entry_point.c)
target_link_libraries(sbcl_librarian PUBLIC sbcl)
```

`sbcl_librarian.c` and `sbcl_librarian.h` are auto-generated bindings for this library as defined in `generate-bindings.lisp`:

```lisp
(define-aggregate-library sbcl-librarian (:function-linkage "LIBSBCL_LIBRARIAN_API")
  diagnostics
  environment
  errors
  handles
  loader)
```

These bindings contain the functions that FASL-packed DLLs call to load themselves into the running Lisp image.

Additionally, we link in `entry_point.c` which defines the `do_initialize_lisp` function that gets called automatically by the system when `libsbcl_librarian.so` is loaded into a process. This function does the following:

1. Get the path to `sbcl_librarian.core`, which is installed in the same directory as the currently executing DLL.
5. Call SBCL's `initialize_lisp` function with the the above core file to initialize the Lisp runtime.
6. Wire up SBCL's lossage handler so that fatal Lisp errors are caught and handled.

## `sbcl_librarian.core`
Built via `generate-bindings.lisp`. Loads SWANK, the `sbcl-librarian` library, sets up some signal-handling hacks, generates C and Python bindings, then saves the core.

## `include/sbcl_librarian.h`
Generated via `generate-bindings.lisp`. Declares functions and values contained in `libsbcl_librarian.so`

## `sbcl_librarian_err.h`
Declares fatal error handling `jmp_buf`s and flags, as well as the global error enum that gets returned by all wrapper functions.

## Python package
Contains several modules that wrap the raw Python bindings:

- `debug.py` wraps functions for enabling/disabling the debugger + backtraces, memory + handle reporting, doing GC, and profiling
- `errors.py` matches the error enum defined in `sbcl_librarian.h` so that errors returned by C functions can be translated into Python exceptions
- `fixtures.py` has a test fixture that checks for handle leaks
- `version.py` reads the package version from VERSION.txt
- `wrapper.py` defines the `LispObject` class for wrapping handles with release logic, exception types corresponding to the errors enums defined in `sbcl_librarian_err.h`, and functions for wrapping raw binding functions with logic to deal with handles and exceptions

### CODE REVIEW

File: examples/libcalc/example.py [hunk 1]
```diff
@@ -1,5 +1,13 @@
+import os
+import platform
+
+if platform.system() == 'Windows':
+    os.add_dll_directory(os.environ.get('LIBSBCL_PATH'))
+    os.add_dll_directory(os.environ.get('MINGW64_PATH'))
```
Comment (jbouwman) [inline examples/libcalc/example.py:6]: May benefit from a note on behavior of `add_dll_directory`

File: lib/entry_point.c [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    char *core_path = malloc(core_path_size);
+
+    snprintf(core_path, core_path_size, "%ssbcl_librarian.core", libsbcl_librarian_dir);
+
+    const char *init_args[] = {"", "--dynamic-space-size", "8192", "--core", core_path, "--noinform", "--no-userinit"};
+
+#ifdef __linux__
+    dlopen("libsbcl.so", RTLD_NOW | RTLD_GLOBAL);
+    dlopen(libsbcl_librarian_path, RTLD_GLOBAL | RTLD_NOW);
+#endif
```
Comment (jbouwman) [inline lib/entry_point.c]: Add note on role in making symbols visible

File: patches/win32-dll-build.patch [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+--- a/src/runtime/GNUmakefile
++++ b/src/runtime/GNUmakefile
+@@ -132,8 +132,8 @@ libsbcl.a: $(OBJS)
+ 	rm -f $@ ; ar rcs $@ $^
+ 
+ PIC_OBJS = $(subst .o,.pic.o,$(OBJS))
+-libsbcl.so: $(PIC_OBJS)
+-	$(CC) -shared -o $@ $^ $(LIBS) $(SOFLAGS)
++libsbcl.dll: $(PIC_OBJS)
++	$(CC) -shared -o $@ $^ $(LIBS) $(SOFLAGS) -Wl,-export-all-symbols -Wl,--whole-archive -lmpfr -Wl,--no-whole-archive mswin64.def
```
Comment (jbouwman) [inline patches/win32-dll-build.patch:32]: Insert note in modified code regarding linkage of log symbols in mpfr; possibly, note about windows sensitivity to file extension

File: lib/sbcl_librarian_err.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
+#pragma once
```
Comment (jbouwman) [inline lib/sbcl_librarian_err.h:1]: top of file comment might be nice due to the amount of cpp going on here

File: src/errors.lisp [hunk 5]
```diff
@@ -1,5 +1,10 @@
 (in-package #:sbcl-librarian)
 
+(define-type :json
+  :c-type "char*"
+  :python-type "c_char_p"
+  :alien-type sb-alien:c-string)
```
Comment (jbouwman) [inline src/errors.lisp:6]: Does anything else in this MR consult this? (Also, might be nice if exhaust-heap were elsewhere: maybe later)

**Review by jbouwman — APPROVED** [review]
Approved, review comments mainly pertain to nonsubstantive documentation of platform quirks

File: lib/entry_point.c (hunk 2 above)
Comment (kartik-s) [reply lib/entry_point.c]: done

File: src/errors.lisp (hunk 5 above)
Comment (kartik-s) [reply src/errors.lisp:6]: No, but it is used downstream

**Review by eliaslfox — APPROVED** [review]
🚀

File: lib/python/src/sbcl_librarian/wrapper.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        return self._handle
+
+    def __eq__(self, other: Any) -> bool:
+        return isinstance(other, LispObject) and sbcl_librarian.raw.lisp_handle_eq(
+            self.handle, other.handle
+        )
+
+    def __del__(self) -> None:
+        if self._release:
+            # https://stackoverflow.com/questions/8590238/unable-to-reference-an-imported-module-in-del
```
Comment (macrologist) [inline lib/python/src/sbcl_librarian/wrapper.py:62]: :+1:

**Review by macrologist — APPROVED** [review]

**Pull request merged**